BrainPortal/app/controllers/tasks_controller.rb

```diff
@@ -771,6 +771,395 @@ def apply_operation(operation, taskids, do_in_spawn = false, options = {})
+        begin
+          Process.setproctitle "ZenodoUpload ID=#{userfile.id} #{idx+1}/#{userfiles.size}"
+          upload_userfile_to_deposit(deposit, userfile)
+          uploaded << "File: #{userfile.name} (ID=#{userfile.id})"
+        rescue => ex
+          errors   << "File: #{userfile.name} (ID=#{userfile.id}): #{ex.class}: #{ex.message}"
+        end
+      end
+
+      # Contact bourreau and get task's out, err and script
+      Process.setproctitle "ZenodoUpload Task Info ID=#{task.id}"
+      task.capture_job_out_err(task.run_number,100_000,100_000) rescue nil # nums are number of lines
+
+      # Captured special data
+      pretty_json_task = JSON.pretty_generate(task.for_api)
+      runtime_info     = JSON.pretty_generate(task.struct_runtime_info)
+      boutiques_path   = task.class.generated_from.descriptor_path rescue nil
+      boutiques_json   = boutiques_path.to_s.ends_with?(".json") ? File.read(boutiques_path) : nil
+      er1 = upload_text_data_to_deposit(deposit, task.script_text,    "main_cbrain_script-#{task.run_id}.sh")
+      er2 = upload_text_data_to_deposit(deposit, task.cluster_stdout, "captured_stdout-#{task.run_id}.log"  )
+      er3 = upload_text_data_to_deposit(deposit, task.cluster_stderr, "captured_stderr-#{task.run_id}.log"  )
+      er4 = upload_text_data_to_deposit(deposit, task.getlog,         "cbrain_log-#{task.run_id}.log"       )
+      er5 = upload_text_data_to_deposit(deposit, pretty_json_task,    "cbrain_task-#{task.run_id}.json"     )
+      er6 = upload_text_data_to_deposit(deposit, runtime_info,        "runtime_info-#{task.run_id}.json"    )
+      er7 = upload_text_data_to_deposit(deposit, boutiques_json,      "boutiques-#{task.run_id}.json"       ) if boutiques_json
+      [ er1, er2, er3, er4, er5, er6, er7 ].each { |er| errors << er if er.present? }
+
+      if errors.present?
+        Message.send_message(current_user,
+          :message_type  => 'error',
+          :header        => "Could not send a file to Zenodo",
+          :description   => "Some errors occurred while sending files to Zenodo",
+          :variable_text => "For #{view_pluralize(errors.size,"file")}:\n" + errors.join("\n")
+        )
+      end
+
+      if uploaded.present?
+        Message.send_message(current_user,
+          :message_type  => 'notice',
+          :header        => "Uploaded files to Zenodo",
+          :description   => "These files were uploaded to Zenodo",
+          :variable_text => "For #{view_pluralize(uploaded.size,"file")}:\n" + uploaded.join("\n")
+        )
+      end
+    end
+  end
+
+  def upload_userfile_to_deposit(deposit, userfile) #:nodoc:
+    is_col = userfile.is_a?(FileCollection)
+
+    # Identify the content
+    userfile.sync_to_cache
+    content_path = userfile.cache_full_path
+    content_name = userfile.name
+    content_path = create_tmp_tar_for_filecollection(userfile) if is_col
+    content_name = userfile.name + ".tar.gz"                   if is_col
+
+    # Upload
```

Comment:
Code duplication from 1105 to 1115 and from 1133 to 1143.

Reply:
Yes, but I'm not sure how to handle refactor it... will look into it.